src/models/huggingface/baichuan.h

```diff
@@ -0,0 +1,538 @@
+    const float base =
+        std::pow(2, -(std::pow(2, -(std::log2(closest_power_of_2) - 3))));
+    const torch::Tensor powers = torch::arange(
+        /*start=*/1, /*end=*/1 + closest_power_of_2, torch::kFloat32);
+    torch::Tensor slopes = torch::pow(base, powers);
+
+    if (closest_power_of_2 != n_heads) {
+      const float extra_base =
+          std::pow(2, -(std::pow(2, -(std::log2(2 * closest_power_of_2) - 3))));
+      const int64_t n_remaining_heads =
+          std::min(closest_power_of_2, n_heads - closest_power_of_2);
+      const torch::Tensor extra_powers =
+          torch::arange(/*start=*/1,
+                        /*end=*/1 + 2 * n_remaining_heads,
+                        /*step=*/2,
+                        torch::kFloat32);
+      const torch::Tensor extra_slopes = torch::pow(extra_base, extra_powers);
+      slopes = torch::cat({slopes, extra_slopes}, /*dim=*/0);
+    }
+    if (parallel_args.world_size() > 1) {
+      slopes = slopes.chunk(/*chunks=*/parallel_args.world_size(),
+                            /*dim=*/0)[parallel_args.rank()];
+    }
+    return slopes;
+  }
+
+ private:
+  BaichuanType baichuan_type_;
+
+  // parameter members, must be registered
+  ColumnParallelLinear W_pack_{nullptr};
+
+  RowParallelLinear o_proj_{nullptr};
+
+  // module members without parameters
+  AttentionWithRoPE atten_rope_{nullptr};
+
+  // module members without parameters
+  AttentionWithAlibi atten_alibi_{nullptr};
+
+  // size for local q, k, v
+  std::vector<int64_t> qkv_sizes_;
+};
+TORCH_MODULE(BaichuanAttention);
+
+class BaichuanDecoderLayerImpl : public torch::nn::Module {
+ public:
+  BaichuanDecoderLayerImpl(const ModelArgs& args,
+                           const QuantArgs& quant_args,
+                           const ParallelArgs& parallel_args,
+                           torch::ScalarType dtype,
+                           const torch::Device& device,
+                           BaichuanType baichuan_type) {
+    // register submodules
+    self_attn_ = register_module(
+        "self_attn",
+        BaichuanAttention(args, quant_args, parallel_args, dtype, device,
+          baichuan_type));
+    mlp_ = register_module(
+        "mlp", BaichuanMLP(args, quant_args, parallel_args, dtype, device));
+
+    input_layernorm_ = register_module(
+        "input_layernorm",
+        RMSNormResidual(args.hidden_size(), args.rms_norm_eps(), dtype, device));
+    post_attention_layernorm_ = register_module(
+        "post_attention_layernorm",
+        RMSNormResidual(args.hidden_size(), args.rms_norm_eps(), dtype, device));
+  }
+
+  torch::Tensor forward(torch::Tensor x,
+                        torch::Tensor positions,
+                        torch::Tensor* residual,
+                        KVCache& kv_cache,
+                        const InputParameters& input_params) {
+    torch::Tensor hidden_states;
+    if (residual == nullptr) {
+      residual = new torch::Tensor();
```

Comment:
sounds you forget to release the memory newed here.
in pytorch, a Tensor is essentially a shared ptr already. if we follow the python style, the code can be rewerite in this way.

```
pass in torch::Tensor& residual

if (!residual.defined()) {
    residual = x;
}

```

Reply:
good catch